test/integration/fake_upstream.h

```diff
@@ -70,6 +70,35 @@ class FakeStream : public Http::StreamDecoder, public Http::StreamCallbacks {
 typedef std::unique_ptr<FakeStream> FakeStreamPtr;
 
 /**
+ * Wraps a raw Network::Connection in a safe way, such that the connection can
+ * be placed in a queue for an arbitrary amount of time. It handles disconnects
+ * that take place in the queued state by failing the test. Once a
+ * QueuedConnectionWrapper object is instantiated by FakeHttpConnection or
+ * FakeRawConnection, it no longer plays a role.
```

Comment:
nit: I would add a TODO here that if we add remove callback support from connection we can simplify this.